doc/CRM-WorkFlow.md

```diff
@@ -0,0 +1,115 @@
+     */
+    sai_status_t sai_object_type_get_availability(
+            _In_ sai_object_id_t switch_id,
+            _In_ sai_object_type_t object_type,
+            _In_ uint32_t attr_count,
+            _In_ const sai_attribute_t *attr_list,
+            _Out_ uint64_t *count);
+```
+__sai_object_type_get_availability()__ API can be used using single parameter object_type or a combination of attributes to create a granular query.
+
+One of the challenges with this API is when the HW resource is shared between more then one such granular query for e.g. Nexthop Group HW table is shared for different types of next hop groups, v4/v6/mpls table is a flexible shared ressource and so on.
+This document is not trying to address shared HW resource but is mainly trying to create a well know workflow.
+
+Also there are legacy read only attributes and can be queried using object specific GET API or can also use the get_availability() API. Recommendation is to migrate NOS to use get_availability() API for consistency reasons. Till then SAI Adapter must support both methods.
+
+# Table of contents
+1. [Query Available System Ports](#introduction)
+2. [Query Available Fabric Ports](#introduction1)
+3. [NextHopGroup and Member Query](#introduction2)
+   1. [Query Available Nexthopgroups](#sub-introduction1)
+   2. [Query Available Nexthopgroups Members](#sub-introduction2)
+4. [Query Available VoQs](#introduction3)
+
+## Query Available System Ports <a name="introduction"></a>
+Query the system port object to get the available system ports. No additional attribute are needed for this query.
+
+>> sai_object_type_get_availability(<switch_id>, SAI_OBJECT_TYPE_SYSTEM_PORT, 0, NULL, &count)
+
+## Query Available Fabric Ports <a name="introduction1"></a>
+Specific port can be queried using the port type attribute. Only one attribure is neded for this query. To query different port types, NOS must invoke the API again with a different port type.
+
+>> sai_attribute_t attr[1];
+>> attr[0].id = SAI_PORT_ATTR_TYPE;
+>> attr[0].value = SAI_PORT_TYPE_FABRIC; // SAI_PORT_TYPE_CPU or SAI_PORT_TYPE_LOOPBACK
+>> sai_object_type_get_availability(<switch_id>, SAI_OBJECT_TYPE_PORT, 1, attr, &count)
+
+## NextHopGroup and Member Query <a name="introduction2"></a>
+SAI supports default as NHG that contains tunnel and IP nexthop members. In this case usually there is a single HW table for NHG. This HW architecture is defined using SAI_NEXT_HOP_GROUP_ATTR_HIERARCHICAL_NEXTHOP with default value as true.
+
+For cases where HW has two different physical tables, one to host Tunnel+IP overlay nexthops and other to host IP only underlay nexthops; SAI_NEXT_HOP_GROUP_ATTR_HIERARCHICAL_NEXTHOP can be quried using capability query API.
+
+Once NOS has determined the HW architecture it can make following queries to determine the available NHG and NHG member entries.
+
+![](./figures/HECMP.png "Figure 1: Hierarchical ECMP")
+__Figure 1: Hierarchical ECMP__
+

```

Comment:
@JaiOCP  If there is a 3-rd level of Nexthop group, how will we able to get the available resource count?